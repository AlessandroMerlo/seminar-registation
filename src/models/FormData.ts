export default interface FormData {
    firstTab: {
        attendeeNumber: string;
        attendeeNames: string[];
    };
    secondTab: {
        isBadges: string | null;
        textInput?: string;
        isAccomodation: string | null;
    };
    thirdTab: {
        isRock: boolean | null;
    };
}