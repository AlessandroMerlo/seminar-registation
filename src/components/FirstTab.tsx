import React, { useState } from 'react';
import styles from './Tab.module.css'
import CheckmarkTab from './CheckmarkTab';
import FormData from '../models/FormData';
import { joinCssClasses } from '../utils/UtilCss';
import { QUESTIONS } from '../utils/Constants';

type Props = {
    data: FormData['firstTab'],
    onUpdate: (data: Partial<FormData['firstTab']>) => void,
    isTabValid: boolean
}

const FirstTab: React.FC<Props> = (prop: Props) => {

    const onChangeAttendeeNumber = (newValue: string): void => {
        prop.onUpdate({
            attendeeNumber: newValue,
            attendeeNames: newValue === '' ? [] : Array(parseInt(newValue)).fill('')
        });
    }

    const onChangeAttendeeName = (index: number, newValue: string): void => {
        const newAttendeeNames = [...prop.data.attendeeNames];
        newAttendeeNames[index] = newValue;
        prop.onUpdate({ attendeeNames: newAttendeeNames });
    }

    return <>
        <div className={joinCssClasses([styles.firstTab, styles.tab])}>
            <div className={styles.headerContainer}>
                <h2 className={joinCssClasses([styles.header, 'helvetica'])}>
                    Step 1
                </h2>
            </div>
            <div className='youbee'>
                <section className={styles.section}>
                    <span>{QUESTIONS.firstTab.firstQuestion}</span>
                    <select
                        onChange={(e) => onChangeAttendeeNumber(e.target.value)}
                        className={styles.attendeeeNumberSelect}
                        value={prop.data.attendeeNumber}
                    >
                        <option value=''>Please Choose</option>
                        {
                            [...Array(5)].map((x, i) =>
                                <option value={`${i + 1}`} key={i + 1}>{i + 1}</option>
                            )
                        }
                    </select>
                    {
                        prop.data.attendeeNames.length > 0 &&
                        <div className={styles.attendeeNameSection}>
                            <p className={joinCssClasses([styles.attendeeNamesParagraph, 'helvetica'])}>
                                Please provide full names:
                            </p>
                            <div className={styles.attendeeNamesContainer}>
                                {
                                    prop.data.attendeeNames.map((x, i) => {
                                        const count = i + 1;
                                        return <React.Fragment key={count}>
                                            <label
                                                htmlFor={`attendeeName-${count}`}
                                                className={styles.inputTextLabel}>
                                                Attendee {count} Name:
                                            </label>
                                            <input
                                                type='text'
                                                id={`attendeeName-${count}`}
                                                value={x}
                                                onChange={(e) => onChangeAttendeeName(i, e.target.value)}
                                                className={styles.attendeeNamesInput}
                                            />
                                        </React.Fragment>
                                    }
                                    )
                                }
                            </div>
                        </ div>
                    }
                    {
                        prop.isTabValid &&
                        <footer>
                            <CheckmarkTab />
                        </footer>
                    }
                </section>
            </div>
        </div>
    </>;
}

export default FirstTab;