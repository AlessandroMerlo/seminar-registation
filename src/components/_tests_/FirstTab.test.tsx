import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import FirstTab from '../FirstTab';
import FormData from '../../models/FormData';
import '@testing-library/jest-dom'

// Mock delle costanti e delle funzioni di utilitÃ 
jest.mock('../../utils/Constants', () => ({
    QUESTIONS: {
        firstTab: {
            firstQuestion: 'How many attendees will be there?',
        },
    },
}));

jest.mock('../../utils/UtilCss', () => ({
    joinCssClasses: (classes: string[]) => classes.join(' '),
}));

jest.mock('../CheckmarkTab', () => () => <div data-testid="checkmark-tab">Checkmark</div>);

// Mock dei dati di esempio
const mockData: FormData['firstTab'] = {
    attendeeNumber: '',
    attendeeNames: [],
};

const mockOnUpdate = jest.fn();

describe('FirstTab Component', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('renders the component with the initial state', () => {
        render(
            <FirstTab data={mockData} onUpdate={mockOnUpdate} isTabValid={false} />
        );

        expect(screen.getByText('Step 1')).toBeInTheDocument();
        expect(screen.getByText('How many attendees will be there?')).toBeInTheDocument();
        expect(screen.getByText('Please Choose')).toBeInTheDocument();
    });

    it('updates attendee number and names correctly', () => {
        render(
            <FirstTab data={mockData} onUpdate={mockOnUpdate} isTabValid={false} />
        );

        const select = screen.getByRole('combobox');
        fireEvent.change(select, { target: { value: '3' } });

        expect(mockOnUpdate).toHaveBeenCalledWith({
            attendeeNumber: '3',
            attendeeNames: ['', '', ''],
        });
    });

    it('updates attendee names correctly', () => {
        const dataWithAttendees = {
            ...mockData,
            attendeeNumber: '2',
            attendeeNames: ['', ''],
        };

        render(
            <FirstTab data={dataWithAttendees} onUpdate={mockOnUpdate} isTabValid={false} />
        );

        const inputs = screen.getAllByRole('textbox');
        fireEvent.change(inputs[0], { target: { value: 'John' } });

        expect(mockOnUpdate).toHaveBeenCalledWith({
            attendeeNames: ['John', ''],
        });
    });

    it('renders the checkmark when isTabValid is true', () => {
        render(
            <FirstTab data={mockData} onUpdate={mockOnUpdate} isTabValid={true} />
        );

        expect(screen.getByTestId('checkmark-tab')).toBeInTheDocument();
    });

    it('matches the snapshot', () => {
        const { asFragment } = render(
            <FirstTab data={mockData} onUpdate={mockOnUpdate} isTabValid={false} />
        );

        expect(asFragment()).toMatchSnapshot();
    });
});