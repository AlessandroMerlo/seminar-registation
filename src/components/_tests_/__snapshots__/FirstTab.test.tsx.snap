// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirstTab Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class=" "
  >
    <div>
      <h2
        class=" helvetica"
      >
        Step 1
      </h2>
    </div>
    <div
      class="youbee"
    >
      <section>
        <span>
          How many attendees will be there?
        </span>
        <select>
          <option
            value=""
          >
            Please Choose
          </option>
          <option
            value="1"
          >
            1
          </option>
          <option
            value="2"
          >
            2
          </option>
          <option
            value="3"
          >
            3
          </option>
          <option
            value="4"
          >
            4
          </option>
          <option
            value="5"
          >
            5
          </option>
        </select>
      </section>
    </div>
  </div>
</DocumentFragment>
`;
